compiler/rustc_mir_dataflow/src/move_paths/builder.rs

```diff
@@ -115,44 +115,56 @@ impl<'b, 'a, 'tcx> Gatherer<'b, 'a, 'tcx> {
-                    ));
-                }
-                ty::Adt(adt, _) if adt.is_union() => {
-                    union_path.get_or_insert(base);
-                }
-                ty::Slice(_) => {
-                    return Err(MoveError::cannot_move_out_of(
-                        self.loc,
-                        InteriorOfSliceOrArray {
-                            ty: place_ty,
-                            is_index: matches!(elem, ProjectionElem::Index(..)),
-                        },
-                    ));
-                }
-
-                ty::Array(..) => {
-                    if let ProjectionElem::Index(..) = elem {
+            match elem {
+                ProjectionElem::Deref => match place_ty.kind() {
+                    ty::Ref(..) | ty::RawPtr(..) => {
                         return Err(MoveError::cannot_move_out_of(
                             self.loc,
-                            InteriorOfSliceOrArray { ty: place_ty, is_index: true },
+                            BorrowedContent {
+                                target_place: place_ref.project_deeper(&[elem], tcx),
+                            },
                         ));
                     }
-                }
-
-                _ => {}
-            };
+                    _ => (),
+                },
+                ProjectionElem::Field(_, _)
+                | ProjectionElem::OpaqueCast(_)
+                | ProjectionElem::Downcast(_, _) => match place_ty.kind() {
+                    ty::Adt(adt, _) if adt.has_dtor(tcx) && !adt.is_box() => {
+                        return Err(MoveError::cannot_move_out_of(
+                            self.loc,
+                            InteriorOfTypeWithDestructor { container_ty: place_ty },
+                        ));
+                    }
+                    ty::Adt(adt, _) if adt.is_union() => {
+                        union_path.get_or_insert(base);
+                    }
 
+                    _ => (),
+                },
+                ProjectionElem::ConstantIndex { .. }
+                | ProjectionElem::Index(_)
+                | ProjectionElem::Subslice { .. } => match place_ty.kind() {
+                    ty::Slice(_) => {
+                        return Err(MoveError::cannot_move_out_of(
+                            self.loc,
+                            InteriorOfSliceOrArray {
+                                ty: place_ty,
+                                is_index: matches!(elem, ProjectionElem::Index(..)),
+                            },
+                        ));
+                    }
+                    ty::Array(..) => {
+                        if let ProjectionElem::Index(..) = elem {
+                            return Err(MoveError::cannot_move_out_of(
+                                self.loc,
+                                InteriorOfSliceOrArray { ty: place_ty, is_index: true },
+                            ));
+                        }
+                    }
```

Comment:
instead of matching on `elem` again here, add a separate branch for `Index` and `ConstantIndex/Subslice` :thinking: given that it changes `is_index` I think that duplication shouldn't negatively impact readability.

Reply:
the inner match is a noop now.

While small mistakes like this can always happen, it may be good to be a bit more attentive when committing changes for review to avoid unnecessary back and forth.

Reply:
sorry about that, will be more attentive next time.